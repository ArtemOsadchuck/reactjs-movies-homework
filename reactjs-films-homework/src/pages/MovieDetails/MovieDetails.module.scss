$background-color: rgb(37, 37, 37);
$grey-color: rgb(121, 121, 121);
$white-color: rgb(255, 255, 255);
$black: rgb(0, 0, 0);

.mainWrapper {
  align-items: center;
  background: $background-color;
  color: $grey-color;
  margin: 0 auto;
  max-width: 1100px;
  min-height: calc(100vh - 5rem);
  position: relative;
  width: 90%;

  .castWrapper {
    align-items: flex-end;
    display: flex;
    flex-direction: column;
    margin-top: 2rem;
    // padding-right: 1.5rem;
    right: 0;
    width: 100%;

    @media screen and (max-width: 1220px) {
      // padding-right: 0;
    }
  }

  .castGrid {
    column-gap: 0.65rem;
    display: flex;
    flex-wrap: wrap;
    grid-gap: 0.65rem;
    padding-right: 1.5rem;

    @media screen and (max-width: 1220px) {
      justify-content: center;
      margin: 0 auto;
      padding: 0 1.5rem;
    }
  }

  .castNameShowAllBtn {
    align-items: center;
    border: 2px solid $white-color;
    border-radius: 5px;
    color: $white-color;
    display: flex;
    flex-shrink: 2;
    font-size: 0.8rem;
    height: 1.8rem;
    justify-content: center;
    transition: 0.3s;
    width: 4.5rem;

    &:hover {
      background-color: $white-color;
      color: $black;
      cursor: pointer;
    }
  }

  .castNameWrapper {
    align-items: flex-start;
    color: $white-color;
    display: flex;
    flex-direction: row;
    height: 2.5rem;
    justify-content: space-between;
    padding-right: 1.5rem;
    width: 68%;

    @media screen and (max-width: 1220px) {
      align-items: center;
      column-gap: 5rem;
      justify-content: space-between;
      margin: 0 auto;
      min-width: 12rem;
      padding: 0 1.5rem;
      width: 48rem;
    }

    @media screen and (max-width: 855px) {
      width: 100%;
    }
  }

  .recommendationsWrapper {
    margin: 0 auto;
    margin-top: 2rem;

    .recommendationsTitle {
      color: $white-color;
      margin-bottom: 2rem;
      padding-left: 1.5rem;
      text-align: start;
    }
  }

  .recommendationsCardsWrapper {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-around;
    width: 100%;

    @media screen and (max-width: 455px) {
      justify-content: center;
    }
  }
}
