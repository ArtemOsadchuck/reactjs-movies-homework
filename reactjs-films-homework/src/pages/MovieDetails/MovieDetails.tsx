import React, { useEffect, useState } from 'react';
import MovieTitleCard from './MovieTitleCard';
import styles from './MovieDetails.module.scss';

import TopBilledCast from './TopBilledCast';
import { ITopBilledCastProp } from './TopBilledCast/TopBilledCast';
import ImagesBlock from './ImagesBlock';
import MovieCard from '../../components/MovieCard';
import { IMovieCard } from '../../components/MovieCard/MovieCard';
import { ITitleMovieProps } from './MovieTitleCard/MovieTitleCard';
import { IImagesBlockProps } from './ImagesBlock/ImagesBlock';

import lang from '../../languages/getLanguage';
import { useAppSelector, useAppDispatch } from '../../hooks/hooks';
import getMovieDetailsData from '../../store/rootStore/movieDetailsPageStore/getMoviePageData/getMovieDetailsData';
import getTopBilletCastData from '../../store/rootStore/movieDetailsPageStore/getMoviePageData/getTopBilletCastData';
import getMovieImages from '../../store/rootStore/movieDetailsPageStore/getMoviePageData/getMovieImages';
import getRecommendations from '../../store/rootStore/movieDetailsPageStore/getMoviePageData/getRecommendations';
import getGenres from '../../store/rootStore/mainStore/getGenres';

const MovieDetails: React.FC = () => {
  const [titleInfoState, setTitleInfoState] = useState<ITitleMovieProps>();
  const [cast, setCast] = useState<Array<ITopBilledCastProp>>();
  const [sortCast, setSortCast] = useState<Array<ITopBilledCastProp>>();
  const [stateImg, setStateImg] = useState<Array<IImagesBlockProps>>();
  const [recommended, setRecommended] = useState<Array<IMovieCard>>();
  // const appFetchMovieGenre = useAppSelector((state) => state.mainReducer.genre);

  const recommendationsQuality = 5;

  const dispatch = useAppDispatch();
  const appLang = useAppSelector((state) => state.mainReducer.lang);
  const ImagesBlockTitle = lang(appLang).images;

  const movie_id = useAppSelector(
    (state) => state.movieDetailsReducer.movie_id
  );
  const appFetchInfo = useAppSelector(
    (state) => state.movieDetailsReducer.moviePageInfoResult
  );
  const recommendations = useAppSelector(
    (state) => state.movieDetailsReducer.recommendations
  );
  const imagesFromAPI = useAppSelector(
    (state) => state.movieDetailsReducer.images
  );
  const movieCastFromAPI = useAppSelector(
    (state) => state.movieDetailsReducer.cast
  );
  useEffect(() => {
    const detailsState = {
      lang: appLang,
      movie_id: movie_id,
    };
    const getData = async () => {
      await dispatch(getMovieDetailsData(detailsState));
      await dispatch(getTopBilletCastData(detailsState));
      await dispatch(getMovieImages(detailsState.movie_id));
      dispatch(getRecommendations(detailsState));
    };

    getData();
  }, [dispatch, appLang, movie_id]);

  useEffect(() => {
    dispatch(getGenres(appLang));
  }, [dispatch, appLang]);

  useEffect(() => {
    setTitleInfoState(() => appFetchInfo);
    setCast(() => movieCastFromAPI?.cast);
    setStateImg(() => imagesFromAPI?.backdrops);
    setRecommended(() => recommendations?.results);
  }, [
    appLang,
    cast,
    appFetchInfo,
    movieCastFromAPI,
    imagesFromAPI,
    recommendations,
  ]);

  const sortCastBySix = (cast: Array<ITopBilledCastProp>) => {
    const castLength = 6;
    const resultCast = [...cast].sort((a, b) => {
      return a.popularity > b.popularity ? -1 : 1;
    });
    return resultCast.slice(0, castLength);
  };

  useEffect(() => {
    cast?.length !== undefined && setSortCast(sortCastBySix(cast));
  }, [cast, appLang, movieCastFromAPI]);

  return (
    <div className={styles.mainWrapper}>
      {titleInfoState && <MovieTitleCard props={titleInfoState} />}
      <div className={styles.castWrapper}>
        <div className={styles.castNameWrapper}>
          <h3 className={styles.castName}>{lang(appLang).topBilledCast}</h3>
          <div className={styles.castNameShowAllBtn}>
            {lang(appLang).showAll}
          </div>
        </div>
        <div className={styles.castGrid}>
          {sortCast
            ? sortCast.map((el: ITopBilledCastProp) => {
                return <TopBilledCast key={el.id} props={el} />;
              })
            : null}
        </div>
        {stateImg ? (
          <ImagesBlock props={stateImg} title={ImagesBlockTitle} />
        ) : null}
      </div>
      <div className={styles.recommendationsWrapper}>
        <h2 className={styles.recommendationsTitle}>
          {lang(appLang).recommendations}
        </h2>
        <div className={styles.recommendationsCardsWrapper}>
          {recommended ? (
            recommended
              .slice(0, recommendationsQuality)
              .map((el: IMovieCard) => {
                return <MovieCard props={el} key={el.id - 0.5} />;
              })
          ) : (
            <h2>Loading...</h2>
          )}
        </div>
      </div>
    </div>
  );
};
export default MovieDetails;
