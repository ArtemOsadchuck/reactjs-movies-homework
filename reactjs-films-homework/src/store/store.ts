import rootStore from './rootStore/rootStore';

export type RootState = ReturnType<typeof rootStore.getState>;
export type AppDispatch = typeof rootStore.dispatch;
export default rootStore;
