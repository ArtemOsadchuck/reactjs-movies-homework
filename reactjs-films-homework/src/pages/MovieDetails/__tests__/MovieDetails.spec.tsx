import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

import MovieDetails from '../MovieDetails';

import { Provider } from 'react-redux';
import store from '../../../store/store';

describe('MovieDetails', () => {
  let fragment: any;

  const history = createMemoryHistory();
  history.push('/');

  beforeEach(() => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router history={history}>
          <MovieDetails />
        </Router>
      </Provider>
    );
    fragment = asFragment();
  });

  afterEach(() => cleanup());

  it('MovieDetails snapshot', () => {
    expect(fragment).toMatchSnapshot();
  });

  it('MovieDetails heading`s must defined', () => {
    expect(screen.getAllByRole('heading')).toBeDefined();
    screen.getAllByRole('heading').forEach((el) => {
      expect(el).toBeDefined();
    });
  });

  it('MovieDetails ShowAll button include class:', () => {
    expect(screen.getByText(/Show|all/i)).toHaveClass('castNameShowAllBtn');
  });

  it('MovieDetails recommendations must be in the document:', () => {
    expect(screen.getByText(/RECOMMENDATIONS/i)).toBeInTheDocument();
  });

  it('MovieDetails Loading... must be in the document:', () => {
    expect(screen.getByText(/Loading.../i)).toBeInTheDocument();
  });
});
