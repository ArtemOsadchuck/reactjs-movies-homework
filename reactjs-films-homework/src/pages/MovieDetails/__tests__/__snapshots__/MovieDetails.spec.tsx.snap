// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails MovieDetails snapshot 1`] = `
<DocumentFragment>
  <div
    class="mainWrapper"
  >
    <div
      class="castWrapper"
    >
      <div
        class="castNameWrapper"
      >
        <h3
          class="castName"
        >
          Top Billed Cast
        </h3>
        <div
          class="castNameShowAllBtn"
        >
          Show all
        </div>
      </div>
      <div
        class="castGrid"
      />
    </div>
    <div
      class="recommendationsWrapper"
    >
      <h2
        class="recommendationsTitle"
      >
        RECOMMENDATIONS
      </h2>
      <div
        class="recommendationsCardsWrapper"
      >
        <h2>
          Loading...
        </h2>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
